hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java

```diff
@@ -326,20 +330,32 @@ private ClientIterator(SqlRowMetadata rowMetadata) {
         }
 
         @Override
-        public boolean hasNext() {
+        public HasNextResult hasNext(long timeout, TimeUnit timeUnit) {
+            long waitNanos = timeUnit.toNanos(timeout);
             while (currentPosition == currentRowCount) {
                 // Reached end of the page. Try fetching the next one if possible.
-                if (!last) {
-                    SqlPage page = fetch();
-
+                if (!last && waitNanos > 0) {
+                    long startNanos = System.nanoTime();
+                    SqlPage page = fetch(waitNanos, NANOSECONDS);
+                    if (page == null) {
+                        return HasNextResult.TIMEOUT;
+                    }
+                    waitNanos -= (System.nanoTime() - startNanos);
                     onNextPage(page);
-                } else {
+                } else if (last) {
                     // No more pages expected, so return false.
-                    return false;
+                    return HasNextResult.DONE;
+                } else {
+                    return HasNextResult.TIMEOUT;
                 }
             }
```

Comment:
We don't need to measure the timeout here, it's already measured in `fetch(timeout, timeUnit)`:
```java
            while (currentPosition == currentRowCount) {
                // Reached end of the page. Try fetching the next one if possible.
                if (last) {
                    return HasNextResult.DONE;
                } else {
                    SqlPage page = fetch(timeout, timeUnit);
                    if (page == null) {
                        return HasNextResult.TIMEOUT;
                    }
                    onNextPage(page);
                }
            }
```
```